From 862fa4324334ba223ea4da1c866f58cac1302bed Mon Sep 17 00:00:00 2001
From: David Bauer <mail@david-bauer.net>
Date: Thu, 6 Aug 2026 01:48:03 +0200
Subject: [PATCH 7/9] gluon-dnsmasq: add package

This is to have a package available for selection which determines
whether Gluon is built with dnsmasq-full or regular dnsmasq.

Signed-off-by: David Bauer <mail@david-bauer.net>
---
 .../luasrc/lib/gluon/upgrade/820-dns-config   | 17 ++++++-
 package/gluon-dnsmasq/Makefile                | 44 +++++++++++++++++++
 package/gluon-setup-mode/Makefile             |  2 +-
 package/gluon-wan-dnsmasq/Makefile            |  2 +-
 4 files changed, 62 insertions(+), 3 deletions(-)
 create mode 100644 package/gluon-dnsmasq/Makefile

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config b/package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config
index 0db66301..c9425e33 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/820-dns-config
@@ -8,11 +8,26 @@ local next_node = site.next_node({})
 
 local dnsmasq = uci:get_first("dhcp", "dnsmasq")
 
+local function dnsmasq_tiny_selected()
+	local f = io.open("/lib/gluon/gluon-dnsmasq-variant", "r")
+	if f then
+		local variant = f:read("*l")
+		f:close()
+		return variant == "tiny"
+	end
+	return nil
+end
+
 uci:set('dhcp', dnsmasq, 'localise_queries', true)
 uci:set('dhcp', dnsmasq, 'localservice', false)
 
 uci:set('dhcp', dnsmasq, 'server', dns.servers)
-uci:set('dhcp', dnsmasq, 'cachesize', dns.cacheentries)
+
+if dnsmasq_tiny_selected() then
+	uci:set('dhcp', dnsmasq, 'cachesize', 0)
+else
+	uci:set('dhcp', dnsmasq, 'cachesize', dns.cacheentries)
+end
 
 uci:delete('firewall', 'client_dns')
 if dns.servers then
diff --git a/package/gluon-dnsmasq/Makefile b/package/gluon-dnsmasq/Makefile
new file mode 100644
index 00000000..f2c6eb72
--- /dev/null
+++ b/package/gluon-dnsmasq/Makefile
@@ -0,0 +1,44 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=gluon-dnsmasq
+
+include ../gluon.mk
+
+define Package/gluon-dnsmasq
+  TITLE:=dnsmasq meta-package
+  DEPENDS:= \
+	+gluon-core \
+	+GLUON_DNSMASQ_VARIANT_TINY:dnsmasq \
+	+GLUON_DNSMASQ_VARIANT_DEFAULT:dnsmasq-full
+endef
+
+define Package/gluon-dnsmasq/description
+	Gluon dnsmasq package selection
+endef
+
+define Package/gluon-dnsmasq/config
+	choice
+		prompt "Dnsmasq variant"
+		default GLUON_DNSMASQ_VARIANT_DEFAULT
+
+	config GLUON_DNSMASQ_VARIANT_TINY
+		bool "Tiny"
+
+	config GLUON_DNSMASQ_VARIANT_DEFAULT
+		bool "Default"
+	endchoice
+endef
+
+define Package/gluon-dnsmasq/install
+	$(Gluon/Build/Install)
+
+	$(INSTALL_DIR) $(1)/lib/gluon
+  ifneq ($(CONFIG_GLUON_DNSMASQ_VARIANT_TINY),y)
+	echo 'tiny' > $(1)/lib/gluon/gluon-dnsmasq-variant
+  else
+	echo 'default' > $(1)/lib/gluon/gluon-dnsmasq-variant
+  endif
+endef
+
+$(eval $(call BuildPackageGluon,gluon-dnsmasq))
+
diff --git a/package/gluon-setup-mode/Makefile b/package/gluon-setup-mode/Makefile
index acc012de..f4266fab 100644
--- a/package/gluon-setup-mode/Makefile
+++ b/package/gluon-setup-mode/Makefile
@@ -9,7 +9,7 @@ include ../gluon.mk
 
 define Package/gluon-setup-mode
   TITLE:=Setup mode
-  DEPENDS:=+gluon-core +gluon-lock-password +ubus +dnsmasq-full \
+  DEPENDS:=+gluon-core +gluon-lock-password +ubus +gluon-dnsmasq \
     +!BUSYBOX_CONFIG_UNSHARE:unshare
 endef
 
diff --git a/package/gluon-wan-dnsmasq/Makefile b/package/gluon-wan-dnsmasq/Makefile
index 44a68f3b..3236f60c 100644
--- a/package/gluon-wan-dnsmasq/Makefile
+++ b/package/gluon-wan-dnsmasq/Makefile
@@ -6,7 +6,7 @@ include ../gluon.mk
 
 define Package/gluon-wan-dnsmasq
   TITLE:=Support for a secondary DNS server using the WAN interface
-  DEPENDS:=+gluon-core +libubus-lua +dnsmasq-full +libpacketmark
+  DEPENDS:=+gluon-core +libubus-lua +gluon-dnsmasq +libpacketmark
 endef
 
 define Package/gluon-wan-dnsmasq/description
-- 
2.53.0

